src/editor.c

```diff
@@ -697,51 +700,111 @@ static void request_reshowing_calltip(SCNotification *nt)
 }
 
 
-static void autocomplete_scope(GeanyEditor *editor)
+static gboolean autocomplete_scope(GeanyEditor *editor, const gchar *root, gsize rootlen)
 {
 	ScintillaObject *sci = editor->sci;
 	gint pos = sci_get_current_position(editor->sci);
 	gchar typed = sci_get_char_at(sci, pos - 1);
+	gchar brace_char;
 	gchar *name;
-	const GPtrArray *tags = NULL;
-	const TMTag *tag;
 	GeanyFiletype *ft = editor->document->file_type;
+	GPtrArray *tags;
+	gboolean function = FALSE;
+	gboolean member;
+	gboolean ret = FALSE;
+	const gchar *current_scope;
 
-	if (ft->id == GEANY_FILETYPES_C || ft->id == GEANY_FILETYPES_CPP)
+	if (autocomplete_scope_shown)
 	{
-		if (pos >= 2 && (match_last_chars(sci, pos, "->") || match_last_chars(sci, pos, "::")))
+		/* move at the operator position */
+		pos -= rootlen;
+
+		/* allow for a space between word and operator */
+		while (pos > 0 && isspace(sci_get_char_at(sci, pos - 1)))
 			pos--;
-		else if (ft->id == GEANY_FILETYPES_CPP && pos >= 3 && match_last_chars(sci, pos, "->*"))
+
+		if (pos > 0)
+			typed = sci_get_char_at(sci, pos - 1);
+	}
+
+	if (typed == '.')
+		pos -= 1;
+	else if (ft->id == GEANY_FILETYPES_C || ft->id == GEANY_FILETYPES_CPP ||
+		ft->id == GEANY_FILETYPES_PHP || ft->id == GEANY_FILETYPES_RUST)
+	{
+		if (match_last_chars(sci, pos, "::"))
```

Comment:
just a thought: what about using `tm_tag_context_separator()` instead of checking for some filetypes?  Not sure it's sensible at all, but right now I see the check above will leave out GLSL, Zephir and PowerShell (not sure if it'd make sense for the latter, but definitely for the former two).